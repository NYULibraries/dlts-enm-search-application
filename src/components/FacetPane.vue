<template>
    <div class="column enm-pane enm-pane-facets is-2">

        <div
            v-show="displayFacetPane"
            class="content">
            <h2 class="enm-pane-heading is-size-5">Limit by Topic</h2>

            <div class="enm-facets-list">
                <div class="enm-topics enm-facets-group-visible">

                    <a
                        v-for="topic in topicFacetItemsAlwaysVisible"
                        :key="topic.name"
                        :id="topic.name"
                        href="#"
                        @click="clickTopicFacetItem">
                        {{ topic.name }}
                        <span class="enm-hitcount">({{ topic.numHits }})</span>
                    </a>

                </div>

                <a
                    v-show="( ! facetPane.showAllTopics ) && ( facetPane.topicsFacetList.length > facetPane.topicsFacetListLimit )"
                    class="listui seemore"
                    href="#"
                    @click="facetPane.showAllTopics = ! facetPane.showAllTopics">
                    <i
                        class="fa fa-angle-double-down"
                        aria-hidden="true"></i>See All
                </a>

                <div
                    v-show="( facetPane.showAllTopics ) && ( facetPane.topicsFacetList.length > facetPane.topicsFacetListLimit )"
                    class="enm-topics enm-facets-group-togglable">

                    <a
                        v-for="topic in topicFacetItemsToggleable"
                        :key="topic.name"
                        :id="topic.name"
                        href="#">
                        @click="clickTopicFacetItem"
                        {{ topic.name }}
                        <span class="enm-hitcount">({{ topic.numHits }})</span>
                    </a>

                </div>

                <a
                    v-show="( facetPane.showAllTopics ) && ( facetPane.topicsFacetList.length > facetPane.topicsFacetListLimit )"
                    class="listui seemore"
                    href="#"
                    @click="facetPane.showAllTopics = ! facetPane.showAllTopics">
                    <i
                        class="fa fa-angle-double-up"
                        aria-hidden="true"></i>See Less
                </a>

            </div>

        </div>

    </div>
</template>

<script>
export default {
    name: 'FacetPane',
};
</script>

<style>
</style>
