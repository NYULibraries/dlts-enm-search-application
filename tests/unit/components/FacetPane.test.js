import FacetPane from '@/components/FacetPane';

import merge from 'lodash.merge';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import Vuex from 'vuex';

const NUM_TOPICS_IN_DEFAULT_TOPICS_FACET_LIST = 25;
const DEFAULT_TOPICS_FACET_LIST = [];
for ( let i = 1; i < NUM_TOPICS_IN_DEFAULT_TOPICS_FACET_LIST; i++ ) {
    DEFAULT_TOPICS_FACET_LIST.push( {
        name    : 'topic' + i,
        numHits : 5 * ( NUM_TOPICS_IN_DEFAULT_TOPICS_FACET_LIST - i ),
    } );
}

const DEFAULT_TOPICS_FACET_LIST_LIMIT = 15;

function createWrapper( overrides ) {
    const defaultMountingOptions = {
        propsData : {
            display              : true,
            topicsFacetList      : DEFAULT_TOPICS_FACET_LIST,
            topicsFacetListLimit : DEFAULT_TOPICS_FACET_LIST_LIMIT,
        },
    };
    return shallowMount( FacetPane, merge( defaultMountingOptions, overrides ) );
}

describe( 'FacetPane', () => {
    describe( 'initialized', () => {
        let wrapper;

        beforeEach( () => {
            wrapper = createWrapper();
        } );

        test( 'has correct showAllTopics', () => {
            expect( wrapper.vm.showAllTopics ).toBeFalsy();
        } );

        test( 'has correct topicFacetItemsAlwaysVisible', () => {
            expect( wrapper.vm.topicFacetItemsAlwaysVisible )
                .toEqual(
                    DEFAULT_TOPICS_FACET_LIST.slice( 0, DEFAULT_TOPICS_FACET_LIST_LIMIT )
                );
        } );

        test( 'has correct topicFacetItemsToggleable', () => {
            expect( wrapper.vm.topicFacetItemsToggleable ).toEqual( [] );
        } );

        test( 'renders correctly', () => {
            expect( wrapper.element ).toMatchSnapshot();
        } );
    } );
    describe( 'click See All', () => {
        let wrapper;

        beforeEach( () => {
            wrapper = createWrapper();

            wrapper.find( '#see-all-link' ).trigger( 'click' );
        } );

        test( 'has correct showAllTopics', () => {
            expect( wrapper.vm.showAllTopics ).toBeTruthy();
        } );

        test( 'has correct topicFacetItemsAlwaysVisible', () => {
            expect( wrapper.vm.topicFacetItemsAlwaysVisible )
                .toEqual(
                    DEFAULT_TOPICS_FACET_LIST.slice( 0, DEFAULT_TOPICS_FACET_LIST_LIMIT )
                );
        } );

        test( 'has correct topicFacetItemsToggleable', () => {
            expect( wrapper.vm.topicFacetItemsToggleable ).toEqual(
                DEFAULT_TOPICS_FACET_LIST.slice( DEFAULT_TOPICS_FACET_LIST_LIMIT )
            );
        } );

        test( 'renders correctly', () => {
            expect( wrapper.element ).toMatchSnapshot();
        } );
    } );
    describe( 'click See Less', () => {
        let wrapper;

        beforeEach( () => {
            wrapper = createWrapper();

            wrapper.find( '#see-less-link' ).trigger( 'click' );
        } );

        test( 'has correct showAllTopics', () => {
            expect( wrapper.vm.showAllTopics ).toBeTruthy();
        } );

        test( 'has correct topicFacetItemsAlwaysVisible', () => {
            expect( wrapper.vm.topicFacetItemsAlwaysVisible )
                .toEqual(
                    DEFAULT_TOPICS_FACET_LIST.slice( 0, DEFAULT_TOPICS_FACET_LIST_LIMIT )
                );
        } );

        test( 'has correct topicFacetItemsToggleable', () => {
            expect( wrapper.vm.topicFacetItemsToggleable ).toEqual(
                DEFAULT_TOPICS_FACET_LIST.slice( DEFAULT_TOPICS_FACET_LIST_LIMIT )
            );
        } );

        test( 'renders correctly', () => {
            expect( wrapper.element ).toMatchSnapshot();
        } );
    } );
    test( 'clickTopicFacetItem calls addSelectedTopicFacetItem action with correct argument', () => {
        let actions;
        let mockAddSelectedTopicFacetItem = jest.fn();
        let store;

        let localVue;
        let wrapper;

        localVue = createLocalVue();

        localVue.use( Vuex );

        actions = {
            addSelectedTopicFacetItem : mockAddSelectedTopicFacetItem,
        };

        store = new Vuex.Store( {
            actions,
        } );

        wrapper = createWrapper( {
            localVue,
            store,
        } );

        const selectedTopicName1 = DEFAULT_TOPICS_FACET_LIST[ 7 ].name;
        const selectedTopicName2 = DEFAULT_TOPICS_FACET_LIST[ 2 ].name;
        const selectedTopicName3 = DEFAULT_TOPICS_FACET_LIST[ 12 ].name;

        wrapper.find( '#' + selectedTopicName1 ).trigger( 'click' );
        wrapper.find( '#' + selectedTopicName2 ).trigger( 'click' );
        wrapper.find( '#' + selectedTopicName3 ).trigger( 'click' );

        expect( mockAddSelectedTopicFacetItem.mock.calls[ 0 ][ 1 ] ).toBe( selectedTopicName1 );
        expect( mockAddSelectedTopicFacetItem.mock.calls[ 1 ][ 1 ] ).toBe( selectedTopicName2 );
        expect( mockAddSelectedTopicFacetItem.mock.calls[ 2 ][ 1 ] ).toBe( selectedTopicName3 );
    } );
    test( 'changing topicFacetList prop updates topics facet list', () => {
        const wrapper = createWrapper();

        const newTopicFacetList = [];
        for ( let i = 1; i < NUM_TOPICS_IN_DEFAULT_TOPICS_FACET_LIST * 2; i++ ) {
            newTopicFacetList.push( {
                name    : 'new-topic-' + i,
                numHits : 88,
            } );
        }

        wrapper.setProps( { topicsFacetList : newTopicFacetList } );

        expect( wrapper.vm.topicFacetItemsAlwaysVisible )
            .toEqual( newTopicFacetList.slice( 0, DEFAULT_TOPICS_FACET_LIST_LIMIT ) );
    } );
    test( 'changing topicFacetListLimit prop updates topics facet list', () => {
        const wrapper = createWrapper();

        const NEW_TOPICS_FACET_LIST_LIMIT = DEFAULT_TOPICS_FACET_LIST_LIMIT + 10;

        wrapper.setProps( { topicsFacetListLimit : NEW_TOPICS_FACET_LIST_LIMIT } );

        expect( wrapper.vm.topicFacetItemsAlwaysVisible )
            .toEqual( DEFAULT_TOPICS_FACET_LIST.slice( 0, NEW_TOPICS_FACET_LIST_LIMIT ) );
    } );
    test( 'setting display prop shows/hides the pane', () => {
        const wrapper = createWrapper();
        const facetsList = wrapper.find( '.enm-facets-list' );

        expect( facetsList.isVisible() ).toBe( true );

        wrapper.setProps( { display : false } );
        expect( facetsList.isVisible() ).toBe( false );

        wrapper.setProps( { display : true } );
        expect( facetsList.isVisible() ).toBe( true );
    } );
} );
