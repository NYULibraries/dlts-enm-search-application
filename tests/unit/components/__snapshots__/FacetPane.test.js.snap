// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FacetPane click See All renders correctly 1`] = `
<div
  class="column enm-pane enm-pane-facets is-2"
>
  <div
    class="content"
  >
    <h2
      class="enm-pane-heading is-size-5"
    >
      
            Limit by Topic
        
    </h2>
     
    <div
      class="enm-facets-list"
    >
      <div
        class="enm-topics enm-facets-group-visible"
      >
        <a
          href="#"
          id="topic1"
        >
          
                    topic1
                    
          <span
            class="enm-hitcount"
          >
            
                        (120)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic2"
        >
          
                    topic2
                    
          <span
            class="enm-hitcount"
          >
            
                        (115)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic3"
        >
          
                    topic3
                    
          <span
            class="enm-hitcount"
          >
            
                        (110)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic4"
        >
          
                    topic4
                    
          <span
            class="enm-hitcount"
          >
            
                        (105)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic5"
        >
          
                    topic5
                    
          <span
            class="enm-hitcount"
          >
            
                        (100)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic6"
        >
          
                    topic6
                    
          <span
            class="enm-hitcount"
          >
            
                        (95)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic7"
        >
          
                    topic7
                    
          <span
            class="enm-hitcount"
          >
            
                        (90)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic8"
        >
          
                    topic8
                    
          <span
            class="enm-hitcount"
          >
            
                        (85)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic9"
        >
          
                    topic9
                    
          <span
            class="enm-hitcount"
          >
            
                        (80)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic10"
        >
          
                    topic10
                    
          <span
            class="enm-hitcount"
          >
            
                        (75)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic11"
        >
          
                    topic11
                    
          <span
            class="enm-hitcount"
          >
            
                        (70)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic12"
        >
          
                    topic12
                    
          <span
            class="enm-hitcount"
          >
            
                        (65)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic13"
        >
          
                    topic13
                    
          <span
            class="enm-hitcount"
          >
            
                        (60)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic14"
        >
          
                    topic14
                    
          <span
            class="enm-hitcount"
          >
            
                        (55)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic15"
        >
          
                    topic15
                    
          <span
            class="enm-hitcount"
          >
            
                        (50)
                    
          </span>
        </a>
      </div>
       
      <a
        class="listui seemore"
        href="#"
        id="see-all-link"
        style="display: none;"
      >
        <i
          aria-hidden="true"
          class="fa fa-angle-double-down"
        />
        See All
            
      </a>
       
      <div
        class="enm-topics enm-facets-group-togglable"
        style=""
      >
        <a
          href="#"
          id="topic16"
        >
          
                    topic16
                    
          <span
            class="enm-hitcount"
          >
            
                        (45)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic17"
        >
          
                    topic17
                    
          <span
            class="enm-hitcount"
          >
            
                        (40)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic18"
        >
          
                    topic18
                    
          <span
            class="enm-hitcount"
          >
            
                        (35)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic19"
        >
          
                    topic19
                    
          <span
            class="enm-hitcount"
          >
            
                        (30)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic20"
        >
          
                    topic20
                    
          <span
            class="enm-hitcount"
          >
            
                        (25)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic21"
        >
          
                    topic21
                    
          <span
            class="enm-hitcount"
          >
            
                        (20)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic22"
        >
          
                    topic22
                    
          <span
            class="enm-hitcount"
          >
            
                        (15)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic23"
        >
          
                    topic23
                    
          <span
            class="enm-hitcount"
          >
            
                        (10)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic24"
        >
          
                    topic24
                    
          <span
            class="enm-hitcount"
          >
            
                        (5)
                    
          </span>
        </a>
      </div>
       
      <a
        class="listui seemore"
        href="#"
        id="see-less-link"
        style=""
      >
        <i
          aria-hidden="true"
          class="fa fa-angle-double-up"
        />
        See Less
            
      </a>
    </div>
  </div>
</div>
`;

exports[`FacetPane click See Less renders correctly 1`] = `
<div
  class="column enm-pane enm-pane-facets is-2"
>
  <div
    class="content"
  >
    <h2
      class="enm-pane-heading is-size-5"
    >
      
            Limit by Topic
        
    </h2>
     
    <div
      class="enm-facets-list"
    >
      <div
        class="enm-topics enm-facets-group-visible"
      >
        <a
          href="#"
          id="topic1"
        >
          
                    topic1
                    
          <span
            class="enm-hitcount"
          >
            
                        (120)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic2"
        >
          
                    topic2
                    
          <span
            class="enm-hitcount"
          >
            
                        (115)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic3"
        >
          
                    topic3
                    
          <span
            class="enm-hitcount"
          >
            
                        (110)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic4"
        >
          
                    topic4
                    
          <span
            class="enm-hitcount"
          >
            
                        (105)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic5"
        >
          
                    topic5
                    
          <span
            class="enm-hitcount"
          >
            
                        (100)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic6"
        >
          
                    topic6
                    
          <span
            class="enm-hitcount"
          >
            
                        (95)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic7"
        >
          
                    topic7
                    
          <span
            class="enm-hitcount"
          >
            
                        (90)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic8"
        >
          
                    topic8
                    
          <span
            class="enm-hitcount"
          >
            
                        (85)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic9"
        >
          
                    topic9
                    
          <span
            class="enm-hitcount"
          >
            
                        (80)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic10"
        >
          
                    topic10
                    
          <span
            class="enm-hitcount"
          >
            
                        (75)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic11"
        >
          
                    topic11
                    
          <span
            class="enm-hitcount"
          >
            
                        (70)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic12"
        >
          
                    topic12
                    
          <span
            class="enm-hitcount"
          >
            
                        (65)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic13"
        >
          
                    topic13
                    
          <span
            class="enm-hitcount"
          >
            
                        (60)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic14"
        >
          
                    topic14
                    
          <span
            class="enm-hitcount"
          >
            
                        (55)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic15"
        >
          
                    topic15
                    
          <span
            class="enm-hitcount"
          >
            
                        (50)
                    
          </span>
        </a>
      </div>
       
      <a
        class="listui seemore"
        href="#"
        id="see-all-link"
        style="display: none;"
      >
        <i
          aria-hidden="true"
          class="fa fa-angle-double-down"
        />
        See All
            
      </a>
       
      <div
        class="enm-topics enm-facets-group-togglable"
        style=""
      >
        <a
          href="#"
          id="topic16"
        >
          
                    topic16
                    
          <span
            class="enm-hitcount"
          >
            
                        (45)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic17"
        >
          
                    topic17
                    
          <span
            class="enm-hitcount"
          >
            
                        (40)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic18"
        >
          
                    topic18
                    
          <span
            class="enm-hitcount"
          >
            
                        (35)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic19"
        >
          
                    topic19
                    
          <span
            class="enm-hitcount"
          >
            
                        (30)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic20"
        >
          
                    topic20
                    
          <span
            class="enm-hitcount"
          >
            
                        (25)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic21"
        >
          
                    topic21
                    
          <span
            class="enm-hitcount"
          >
            
                        (20)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic22"
        >
          
                    topic22
                    
          <span
            class="enm-hitcount"
          >
            
                        (15)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic23"
        >
          
                    topic23
                    
          <span
            class="enm-hitcount"
          >
            
                        (10)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic24"
        >
          
                    topic24
                    
          <span
            class="enm-hitcount"
          >
            
                        (5)
                    
          </span>
        </a>
      </div>
       
      <a
        class="listui seemore"
        href="#"
        id="see-less-link"
        style=""
      >
        <i
          aria-hidden="true"
          class="fa fa-angle-double-up"
        />
        See Less
            
      </a>
    </div>
  </div>
</div>
`;

exports[`FacetPane initialized renders correctly 1`] = `
<div
  class="column enm-pane enm-pane-facets is-2"
>
  <div
    class="content"
  >
    <h2
      class="enm-pane-heading is-size-5"
    >
      
            Limit by Topic
        
    </h2>
     
    <div
      class="enm-facets-list"
    >
      <div
        class="enm-topics enm-facets-group-visible"
      >
        <a
          href="#"
          id="topic1"
        >
          
                    topic1
                    
          <span
            class="enm-hitcount"
          >
            
                        (120)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic2"
        >
          
                    topic2
                    
          <span
            class="enm-hitcount"
          >
            
                        (115)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic3"
        >
          
                    topic3
                    
          <span
            class="enm-hitcount"
          >
            
                        (110)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic4"
        >
          
                    topic4
                    
          <span
            class="enm-hitcount"
          >
            
                        (105)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic5"
        >
          
                    topic5
                    
          <span
            class="enm-hitcount"
          >
            
                        (100)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic6"
        >
          
                    topic6
                    
          <span
            class="enm-hitcount"
          >
            
                        (95)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic7"
        >
          
                    topic7
                    
          <span
            class="enm-hitcount"
          >
            
                        (90)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic8"
        >
          
                    topic8
                    
          <span
            class="enm-hitcount"
          >
            
                        (85)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic9"
        >
          
                    topic9
                    
          <span
            class="enm-hitcount"
          >
            
                        (80)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic10"
        >
          
                    topic10
                    
          <span
            class="enm-hitcount"
          >
            
                        (75)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic11"
        >
          
                    topic11
                    
          <span
            class="enm-hitcount"
          >
            
                        (70)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic12"
        >
          
                    topic12
                    
          <span
            class="enm-hitcount"
          >
            
                        (65)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic13"
        >
          
                    topic13
                    
          <span
            class="enm-hitcount"
          >
            
                        (60)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic14"
        >
          
                    topic14
                    
          <span
            class="enm-hitcount"
          >
            
                        (55)
                    
          </span>
        </a>
        <a
          href="#"
          id="topic15"
        >
          
                    topic15
                    
          <span
            class="enm-hitcount"
          >
            
                        (50)
                    
          </span>
        </a>
      </div>
       
      <a
        class="listui seemore"
        href="#"
        id="see-all-link"
      >
        <i
          aria-hidden="true"
          class="fa fa-angle-double-down"
        />
        See All
            
      </a>
       
      <div
        class="enm-topics enm-facets-group-togglable"
        style="display: none;"
      />
       
      <a
        class="listui seemore"
        href="#"
        id="see-less-link"
        style="display: none;"
      >
        <i
          aria-hidden="true"
          class="fa fa-angle-double-up"
        />
        See Less
            
      </a>
    </div>
  </div>
</div>
`;
